lxd/network/driver_ovn.go

```diff
@@ -214,6 +216,78 @@ func (n *ovn) projectRestrictedSubnets(p *api.Project, uplinkNetworkName string)
 	return projectRestrictedSubnets, nil
 }
 
+func (n *ovn) randomExternalAddress(ctx context.Context, ipVersion int, uplinkRoutes []*net.IPNet, projectRestrictedSubnets []*net.IPNet, validator func(*net.IPNet) error) (*net.IPNet, error) {
+	// Ensure a sensible deadline is set.
+	_, hasDeadline := ctx.Deadline()
+	var cancel context.CancelFunc = func() {}
+	if !hasDeadline {
+		ctx, cancel = context.WithTimeout(ctx, 5*time.Second)
+	}
+
+	defer cancel()
+
+	var subnets []*net.IPNet
+	for _, projectRestrictedSubnet := range projectRestrictedSubnets {
+		if (projectRestrictedSubnet.IP.To4() != nil && ipVersion == 4) || (projectRestrictedSubnet.IP.To16() != nil && ipVersion == 6) {
+			subnets = append(subnets, projectRestrictedSubnet)
+		}
+	}
+
+	if projectRestrictedSubnets == nil {
+		for _, uplinkRoute := range uplinkRoutes {
+			if (uplinkRoute.IP.To4() != nil && ipVersion == 4) || (uplinkRoute.IP.To16() != nil && ipVersion == 6) {
+				subnets = append(subnets, uplinkRoute)
+			}
+		}
+	}
+
+	if len(subnets) == 0 {
+		return nil, fmt.Errorf("No IPv%d routes are configured for this network", ipVersion)
+	}
+
+	if len(subnets) > 1 {
+		// Shuffle the subnets so we aren't always picking from the same one.
+		for i := range subnets {
+			j := rand.Intn(i + 1)
+			subnets[i], subnets[j] = subnets[j], subnets[i]
+		}
+	}
+
+	// Distribute the deadline across the number of subnets we're attempting to find an address in.
+	deadline, _ := ctx.Deadline()
+	perSubnetTimeout := time.Until(deadline) / time.Duration(len(subnets))
+
+	for _, subnet := range subnets {
+		subnetCtx, subnetCtxCancel := context.WithTimeout(ctx, perSubnetTimeout)
+		addressInSubnet, err := randomAddressInSubnet(subnetCtx, subnet, func(n net.IP) error {
+			if validator == nil {
+				return nil
+			}
+
+			ipnet, err := ParseIPToNet(n.String())
+			if err != nil {
+				return err
+			}
+
+			return validator(ipnet)
+		})
+
+		subnetCtxCancel()
+
+		if err != nil {
+			if !errors.Is(err, context.DeadlineExceeded) {
```

Comment:
What if context.DeadlineExceeded comes from the parent context? We shouldnt go onto the next subnet in that case right?

Reply:
I suggest checking ctx.Err() != nil here instead.